File: java/core/src/main/java/com/github/pukkaone/mullet/html/FilteredElementHandler.java
/*
Copyright (c) 2011, Chin Huang
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.github.pukkaone.mullet.html;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.ext.DefaultHandler2;

/**
 * This is a proxy for a SAX event handler that forwards events only while the
 * parser is within an element identified by an {@code id} attribute.
 */
class FilteredElementHandler extends DefaultHandler2 {

    private static final String ID = "id";

    private DefaultHandler2 handler;
    private String id;
    private int depth;

    FilteredElementHandler(DefaultHandler2 handler, String id) {
        this.handler = handler;
        this.id = id;
    }

    private boolean shouldForward() {
        return depth > 0;
    }

    @Override
    public void startDocument() throws SAXException {
        handler.startDocument();
    }

    @Override
    public void endDocument() throws SAXException {
        handler.endDocument();
    }

    @Override
    public void startElement(
            String namespaceUri,
            String localName,
            String qualifiedName,
            Attributes attributes)
        throws SAXException
    {
        if (shouldForward()) {
            ++depth;
            handler.startElement(
                    namespaceUri, localName, qualifiedName, attributes);
            return;
        }

        String value = attributes.getValue(ID);
        if (id.equals(value)) {
            // Enable event forwarding.
            depth = 1;
        }
    }

    @Override
    public void endElement(
            String namespaceUri, String localName, String qualifiedName)
        throws SAXException
    {
        if (shouldForward()) {
            if (--depth > 0) {
                handler.endElement(namespaceUri, localName, qualifiedName);
            }
        }
    }

    @Override
    public void characters(char[] ch, int start, int length)
        throws SAXException
    {
        if (shouldForward()) {
            handler.characters(ch, start, length);
        }
    }

    @Override
    public void ignorableWhitespace(char[] ch, int start, int length)
        throws SAXException
    {
        if (shouldForward()) {
            handler.ignorableWhitespace(ch, start, length);
        }
    }

    @Override
    public void startCDATA() throws SAXException {
        if (shouldForward()) {
            handler.startCDATA();
        }
    }

    @Override
    public void endCDATA() throws SAXException {
        if (shouldForward()) {
            handler.endCDATA();
        }
    }

    @Override
    public void comment(char[] ch, int start, int length)
        throws SAXException
    {
        if (shouldForward()) {
            handler.comment(ch, start, length);
        }
    }

    @Override
    public void processingInstruction(String target, String data)
        throws SAXException
    {
        if (shouldForward()) {
            handler.processingInstruction(target, data);
        }
    }
}

File: ruby/core/lib/mullet/html/filtered_element_handler.rb
require 'mullet/html/command'
require 'mullet/template_error'
require 'nokogiri'
require 'set'

module Mullet; module HTML

  # Proxy for a SAX event handler that forwards events only while the
  # parser is within an element identified by an `id` attribute.
  class FilteredElementHandler < Nokogiri::XML::SAX::Document
    ID = 'id'

    # Constructor
    #
    # @param [Document] handler
    #           event handler to forward events to
    # @param [String] id
    #           id attribute value
    def initialize(handler, id)
      @handler = handler
      @id = id
      @depth = 0
      @found_id = false
    end

    def should_forward()
      return @depth > 0
    end

    def start_document()
      @handler.start_document()
    end

    def end_document()
      @handler.end_document()

      if !@found_id
        raise TemplateError.new("element with attribute id='#{@id}' not found")
      end
    end

    def start_element_namespace(name, attributes, prefix, uri, namespaces)
      if should_forward()
        @depth += 1
        return @handler.start_element_namespace(
            name, attributes, prefix, uri, namespaces)
      end

      value = attributes.each do |attr|
        qualified_name = [attr.prefix, attr.localname].compact.join(':')
        if qualified_name == ID && attr.value == @id
          # Enable event forwarding.
          @depth = 1
          @found_id = true
        end
      end
    end

    def end_element_namespace(name, prefix, uri)
      if should_forward()
        @depth -= 1
        if @depth > 0
          @handler.end_element_namespace(name, prefix, uri)
        end
      end
    end

    def characters(data)
      if should_forward()
        @handler.characters(data)
      end
    end

    def cdata_block(data)
      if should_forward()
        @handler.cdata_block(data)
      end
    end

    def comment(data)
      if should_forward()
        @handler.comment(data)
      end
    end
  end

end; end
